### PR TITLE
[MRG] Refactor standard deviation computation into _return_std helper

### PR DESCRIPTION
**Pull request by MechCoder**

### CODE REVIEW

**Comment by MechCoder** [comment]
@betatim @glouppe

File: skopt/learning/forest.py [hunk 1]
```diff
@@ -3,6 +3,44 @@
 from sklearn.ensemble import ExtraTreesRegressor as sk_ExtraTreesRegressor
 
 
+def _return_std(X, trees, predictions):
+    """
+    Return standard deviation in predictions given a list of trees.
```
Comment (glouppe) [inline skopt/learning/forest.py]: Looks good, except for this docstring. We are computing `std(y|x)`, not `std(\hat(y)|x)`.

Comment (MechCoder) [inline skopt/learning/forest.py]: By given a list of trees, I had meant `given a list of trees as a parameter in the programming sense` and not `std given a tree in the probabilistic sense` . My bad, that it implied that :P

**Comment by MechCoder** [comment]
will merge when travis passes

File: skopt/learning/forest.py (hunk 1 above)
Comment (glouppe) [inline skopt/learning/forest.py]: I was rather saying that we are computing the standard deviation of the target `y`, rather than the standard deviation of the prediction `\hat(y)` of the target. For me "standard deviation in predictions"  meant the later.

File: skopt/learning/forest.py [hunk 2]
```diff
@@ -3,6 +3,47 @@
 from sklearn.ensemble import ExtraTreesRegressor as sk_ExtraTreesRegressor
 
 
+def _return_std(X, trees, predictions):
+    """
+    Returns ``std(Y | X)``.
```
Comment (betatim) [inline skopt/learning/forest.py]: only one ` needed

**Comment by MechCoder** [comment]
@glouppe I admit, I don't completely understand the argument.

The method computes the variance in predictions of each random forest, while giving an estimate of the variance in the target. In other words, we are assuming `std(Y_hat | x)` as an estimate of `std(Y | x)`. The method still computes `std(Y_hat | x) right?`

File: skopt/learning/forest.py (hunk 2 above)
Comment (MechCoder) [inline skopt/learning/forest.py]: Great, I'll push this directly.

**Comment by glouppe** [comment]
> In other words, we are assuming std(Y_hat | x) as an estimate of std(Y | x). The method still computes std(Y_hat | x) right?

No, we are not. Y_hat is itself a random variable, with a corresponding mean and variance, and is distinct from Y. For this precise reason, this is why we dont use `np.std([tree.predict(X) for tree in forest], axis=0)` as an estimate of `std(Y|X)`. This would indeed rather compute the standard deviation `std(Y_hat|X)` of the prediction Y_hat,  over the distribution of random trees.

**Comment by MechCoder** [comment]
Just to make sure we are on the same page, in your explanation `Y_hat` is a random variable that denotes the prediction in each tree?

**Comment by glouppe** [comment]
Yes, Y_hat(_tree) is the prediction of my model (here an individual tree). Note that Y_hat_forest would itself be another random variable, which is distinct from Y_hat(_tree).

**Comment by MechCoder** [comment]
Yes, by "variance in predictions", I was meaning the "variance in predictions of the random forest model" not `Var(Y_hat | X)`. More precisely, the one described as `Var(L)` described in page 17 of the paper. Does it make sense then?

**Comment by glouppe** [comment]
Well, for the same reasons, I understand "variance in predictions of the random forest model"  as `Var(Y_hat_forest | X)`, which for me is distinct from `Var(Y | X)`.

**Comment by MechCoder** [comment]
Ok, it is just a bit confusing to me to say that it "computes" `Var(Y | X)` when it estimates `Var(Y | X)`. The same way that you cannot compute the true mean given data sampled from a distribution (unless you know the distribution) but you can just estimate the mean.

**Comment by glouppe** [comment]
Yes sure, we only approximates it, but still we need to approximate the right quantity.

**Comment by betatim** [comment]
Worth adding some of these definitions/clarifications/shared vocabulary to #23 (and eventually build that issue into a section of the docs)? I can see this issue coming up again and again.

**Comment by MechCoder** [comment]
It is due to my misunderstanding.

For me, by "variance in predictions", I was referring to the variance in the leaf node for each tree but the correct terminology is as described by @glouppe over here (https://github.com/scikit-optimize/scikit-optimize/pull/93#issuecomment-232270233). That is why random forests are used, to reduce the "variance in predictions" :-)

**Comment by MechCoder** [comment]
And I do think, they should be added (preferably by me)
